### PR TITLE
feat(sdk): implement injection

### PR DESCRIPTION
**Pull request by afmika**

### CODE REVIEW

File: typegraph/core/wit/typegraph.wit [hunk 1]
```diff
@@ -15,12 +15,14 @@ interface core {
     finalize-typegraph: func() -> result<string, error>
 
     type type-id = u32
-
     record type-base {
         name: option<string>,
+        // json
+        injection: option<string>,
```
Comment (zifeo) [inline typegraph/core/wit/typegraph.wit:21]: can we type with an enum the different injection?

File: typegraph/deno/src/effects.ts [hunk 2]
```diff
@@ -23,3 +23,9 @@ export function delete_(idempotent = true): EffectDelete {
 export function update(idempotent = true): EffectUpdate {
   return { tag: "update", val: idempotent };
 }
+
+export const effectPrefix = "__effect_";
```
Comment (zifeo) [inline typegraph/deno/src/effects.ts:27]: I believe we shall support simple string as well so I would avoid any prefix

File: typegraph/python_next/typegraph_next/t.py [hunk 3]
```diff
@@ -77,6 +78,66 @@ def optional(self, default_value: Optional[str] = None) -> "optional":
...
+                    isinstance(k, EffectType) for k in value.keys()
+                ):
+                    correct_value = {}
+                    for k, v in value.items():
+                        if not isinstance(v, str):
+                            raise Exception(
+                                f"value for field {k.name} must be a string"
+                            )
+                        correct_value[k] = proxy(v).id
+                else:
```
Comment (zifeo) [inline typegraph/python_next/typegraph_next/t.py:125]: same here, heavily nested

File: typegraph/core/wit/typegraph.wit (hunk 1 above)
Comment (afmika) [reply typegraph/core/wit/typegraph.wit:21]: wit enums cannot have type payloads like in rust.

File: typegraph/deno/src/effects.ts (hunk 2 above)
Comment (afmika) [reply typegraph/deno/src/effects.ts:27]: If we allow simple string then we have to reserve `update`, `delete`, `create`, and `none` for the typescript frontend. 
We can however have a workaround by making `PerEffect` explicit
```typescript
struct({ a: t.string() })
.set({ [NONE]: {a: 1}, [CREATE]: {a: 2} })  // prefix is used to distinguish between an object and a per effect object
```
becomes
```typescript
struct({ a: t.string() })
.set({ none: {a: 1}, create: {a: 2} }, "PerEffect")
```

Comment (Natoandro) [reply typegraph/deno/src/effects.ts:27]: We can use [symbols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol).

File: typegraph/core/wit/typegraph.wit (hunk 1 above)
Comment (Natoandro) [reply typegraph/core/wit/typegraph.wit:21]: But WIT have `variant`, see [WIT types](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md#wit-types).

File: typegraph/core/wit/typegraph.wit [hunk 4]
```diff
@@ -15,12 +15,14 @@ interface core {
...
         // string => json string
         runtime-config: option<list<tuple<string, string>>>
     }
+    update-type-injection: func(id: u32, value: string) -> result<_, error>
```
Comment (Natoandro) [inline typegraph/core/wit/typegraph.wit:25]: I would rather not modify the type, as the same type might be used elsewhere without the injection. See the approach I used for policies.

```
record type-with-injection {
   tpe: type-id,
   injection: ...
}
```

**Review by Natoandro — CHANGES_REQUESTED** [review]
Some suggestions.

File: typegraph/deno/src/utils/func_utils.ts [hunk 5]
```diff
@@ -1,6 +1,63 @@
...
+    value !== null
+  ) {
+    // Note:
+    // Symbol changes the behavior of keys, values, entries => props are skipped
+    const symbols = [UPDATE, DELETE, CREATE, NONE];
+    const noOtherType = Object.keys(value).length == 0;
+    const isPerEffect = noOtherType &&
+      symbols
+        .map((symbol) => (value as any)?.[symbol] !== undefined)
+        .reduce(
```
Comment (Natoandro) [inline typegraph/deno/src/utils/func_utils.ts:32]: [Array.some()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some)

Comment (Natoandro) [inline typegraph/deno/src/utils/func_utils.ts:32]: [`Array.some()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) ?

**Review by Natoandro — APPROVED** [review]
Nice :+1:

Some small suggestion.

**Review by Natoandro — APPROVED** [review]
Nice :+1:

Some small suggestion.

File: typegraph/deno/src/utils/func_utils.ts (hunk 5 above)
Comment (afmika) [reply typegraph/deno/src/utils/func_utils.ts:32]: Indeed this should make it shorter/cleaner.

**Review by zifeo — APPROVED** [review]

**Pull request merged**